### PR TITLE
Sprint20 fix geom multiplot

### LINKED ISSUE #266: Fix plotting multiple samples (samples.plot) when StepExpantion and KLExpantion geometries are used (encountered in the paper example)
@jakobsj, @nabriis 

Just wanted to keep you posted about updating this issue status to ready :). We talked about it today as needed for the heat_1D paper example.

### PR DESCRIPTION
**Pull request by amal-ghamdi**
closes #266

### CODE REVIEW

**Review by jakobsj — COMMENTED** [review]
Hi @amal-ghamdi, many thanks for adding this PR.  I took a look through it -  I'm afraid I couldn't quite remember the problem it is seeking to address, at the same time it is somewhat sizable and technical, the issue and PR do not contain much description, and there isn't a demo or similar so it was a bit difficult for me to process it.  Could I ask you to provide a bit more description of what the problem is and what has been implemented in the PR to address it please?  I think in general we should seek to do this more consistently, at least once the issue/PR is above a certain size. Thanks!

**Comment by amal-ghamdi** [comment]
> Hi @amal-ghamdi, many thanks for adding this PR. I took a look through it - I'm afraid I couldn't quite remember the problem it is seeking to address, at the same time it is somewhat sizable and technical, the issue and PR do not contain much description, and there isn't a demo or similar so it was a bit difficult for me to process it. Could I ask you to provide a bit more description of what the problem is and what has been implemented in the PR to address it please? I think in general we should seek to do this more consistently, at least once the issue/PR is above a certain size. Thanks!

Thank you @jakobsj for taking a look and for your feedback. Here is some more context:

This PR fix the following code (the following code passes when Ns=1 but fails when Ns=2 because par2fun is unable to operate on multiple parameter vectors, e.g. multiple samples). The code example here is for `KLExpansion` geometry, and the same issues arise when using `StepExpansion` as well.
```python
from cuqi.distribution import Gaussian
from cuqi.geometry import KLExpansion
import numpy as np
import matplotlib.pyplot as plt
Ns = 2
grid = np.linspace(0, 1)
kl_expansion = KLExpansion(grid)

y = Gaussian(np.zeros(kl_expansion.par_dim), 1, geometry=kl_expansion)
samples_kl = y.sample(Ns)
plt.figure()
samples_kl.plot()
```

Specific changes:
- Shape of input to `par2fun` and `fun2par` and in these two geometries is processed by two parent methods: ` _reshape_par2fun_input`  and `_reshape_fun2par_input`. 
-  `fun2par` and `par2fun` in these two geometries are generalized to handle multiple functions/ parameter vectors (e.g. multiple samples)
- Some old tests were updated to compare results for fun2par and par2fun with the pre PR implementation (results of the pre PR implementation are added as data files in the tests directory)
- Additional tests added to compare converting multiple (parameter vectors or functions) at once and individually. 

we can discuss more in the stand up

**Comment by amal-ghamdi** [comment]
> > Hi @amal-ghamdi, many thanks for adding this PR. I took a look through it - I'm afraid I couldn't quite remember the problem it is seeking to address, at the same time it is somewhat sizable and technical, the issue and PR do not contain much description, and there isn't a demo or similar so it was a bit difficult for me to process it. Could I ask you to provide a bit more description of what the problem is and what has been implemented in the PR to address it please? I think in general we should seek to do this more consistently, at least once the issue/PR is above a certain size. Thanks!
> 
> Thank you @jakobsj for taking a look and for your feedback. Here is some more context:
> 
> This PR fix the following code (the following code passes when Ns=1 but fails when Ns=2 because par2fun is unable to operate on multiple parameter vectors, e.g. multiple samples). The code example here is for `KLExpansion` geometry, and the same issues arise when using `StepExpansion` as well.
> 
> ```python
> from cuqi.distribution import Gaussian
> from cuqi.geometry import KLExpansion
> import numpy as np
> import matplotlib.pyplot as plt
> Ns = 2
> grid = np.linspace(0, 1)
> kl_expansion = KLExpansion(grid)
> 
> y = Gaussian(np.zeros(kl_expansion.par_dim), 1, geometry=kl_expansion)
> samples_kl = y.sample(Ns)
> plt.figure()
> samples_kl.plot()
> ```
> 
> Specific changes:
> 
> * Shape of input to `par2fun` and `fun2par` and in these two geometries is processed by two parent methods: ` _reshape_par2fun_input`  and `_reshape_fun2par_input`.
> * `fun2par` and `par2fun` in these two geometries are generalized to handle multiple functions/ parameter vectors (e.g. multiple samples)
> * Some old tests were updated to compare results for fun2par and par2fun with the pre PR implementation (results of the pre PR implementation are added as data files in the tests directory)
> * Additional tests added to compare converting multiple (parameter vectors or functions) at once and individually.
> 
> we can discuss more in the stand up

Note: I added two issues for possible future update for `KLExpansion_Full` and `CusomKL`. (#272, #273)

**Review by nabriis — APPROVED** [review]

File: tests/test_geometry.py [hunk 1]
```diff
@@ -458,3 +483,68 @@ def test_Continuous2D_plot_multiple_funvals_pass(geom2D_funvals):
...
+    decay_rate = 2.5
+    normalizer = 12
+    geom = cuqi.geometry.KLExpansion(grid,
+                                     decay_rate=decay_rate,
+                                     normalizer=normalizer,
+                                     num_modes=10)
+    return geom
+
+# Create fixture for StepExpansion geometry
+@pytest.fixture
```
Comment (nabriis) [inline tests/test_geometry.py:502]: Nice with the fixtures :)

**Review by nabriis — APPROVED** [review]
Hi @amal-ghamdi. These changes LGTM. I did not have many comments relating the changes as they are rather specific to these geometries. Good to get fun2par added to KL.

**Review by jakobsj — APPROVED** [review]
Many thanks for adding more context and detailing the changes, that is very helpful. I am happy with the changes, so please go ahead and merge :)

**Pull request merged**